### ISSUE TITLE
How to specify continuous set of uncertain parameters per each discrete scenario in PyROS?  

### ISSUE THREAD

**Issue by makansij** [Pyomo/pyomo#2431]
## Summary

My suggestion is to allow continuous set of uncertain parameters per each discrete scenario

### Rationale

Many practical problems require it, and plus it is still within the GRCS algorithm formulation.  See equation (3).

### Description

For each scenario (each realization of the discrete parameters), allow the user to specify an uncertainty set for the uncertain continuous parameters.  

### Additional information
There are no examples in the online documentation involving Discrete Uncertainty sets. So, I've had to read the source code to figure out how to use it.  It might make it easier for users to understand if there were an example in the docs under "Solve with PyROS".

Corresponding SO  question here: https://stackoverflow.com/questions/72620730

**Comment by blnicho** [comment]
@shermanjasonaf could you comment on this?

**Comment by shermanjasonaf** [comment]
If I understand correctly, are you asking about specifying a different type of constraint for each uncertain parameter (i.e. each dimension of the set)? That is, are you asking about constructing a Cartesian product of multiple uncertainty sets? If this is the case, then we are currently thinking about adding a `CartesianProductSet` class in the future, but there are currently issues with addressing the possibilities of constructing sets with mixed-integer variables/constraints.

Otherwise, can you clarify what you mean by specifying a "continuous set for each scenario", and clarify your reference to equation (3)?

Alternatively, per the [documentation](https://pyomo.readthedocs.io/en/stable/contributed_packages/pyros.html?highlight=pyros#pyros-uncertainty-sets), you may implement a custom uncertainty set by defining your own subclass of the `UncertaintySet` [class](https://pyomo.readthedocs.io/en/stable/contributed_packages/pyros.html?highlight=pyros#pyomo.contrib.pyros.uncertainty_sets.UncertaintySet).

**Comment by makansij** [comment]
@shermanjasonaf YES!  A CartesianProductSet is exactly what I had in mind – I would have used that language but I wasn’t sure if it would get interpreted correctly.   

How can I help with adding this CartesianProductSet class?  Without it,  I’m afraid PyROS cannot faithfully represent the problem formulation in equation (3) of this paper: 
![Screen Shot 2022-06-15 at 7 09 43 AM](https://user-images.githubusercontent.com/8031047/173848544-ad8cab8f-6fa4-4ad7-871c-02006d23443b.png)
https://aiche.onlinelibrary.wiley.com/doi/epdf/10.1002/aic.17175

**Comment by makansij** [comment]
> If I understand correctly, are you asking about specifying a different type of constraint for each uncertain parameter (i.e. each dimension of the set)? That is, are you asking about constructing a Cartesian product of multiple uncertainty sets? If this is the case, then we are currently thinking about adding a `CartesianProductSet` class in the future, but there are currently issues with addressing the possibilities of constructing sets with mixed-integer variables/constraints.
> 
> Otherwise, can you clarify what you mean by specifying a "continuous set for each scenario", and clarify your reference to equation (3)?
> 
> Alternatively, per the [documentation](https://pyomo.readthedocs.io/en/stable/contributed_packages/pyros.html?highlight=pyros#pyros-uncertainty-sets), you may implement a custom uncertainty set by defining your own subclass of the `UncertaintySet` [class](https://pyomo.readthedocs.io/en/stable/contributed_packages/pyros.html?highlight=pyros#pyomo.contrib.pyros.uncertainty_sets.UncertaintySet).

@shermanjasonaf I'll try the custom Uncertainty Set, and see if I can solve it that way.  The documentation seems to indicate that this should be able to support my use case, do you agree?

**Comment by shermanjasonaf** [comment]
You should be able to create an `UncertaintySet` with either of the following geometries: linear, convex, non-convex, or purely discrete. PyROS currently solves separation problems for purely discrete geometries differently than it does for all other geometries. Note that there is currently no support for sets with both discrete and continuous uncertain parameters; this is currently a work in-progress (adding support for such sets will entail writing a separation problem routine to handle these geometries). To this end, you can implement a custom-written Cartesian product set---so long as the arguments of the Cartesian product are all continuous, or all discrete.
